### PR TITLE
Adding Javadoc in ServerGraph to the getFlows(Path) method.

### PR DESCRIPTION
**Pull request by alexscheffler**
Please check the Exception description.

### CODE REVIEW

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 1]
```diff
@@ -741,6 +741,14 @@ public Flow getFlow(int id) throws Exception {
 		}
 	}
 
+	/**
+	 * Finds all the flows that cross at least one server on the path p.
+	 * @param p
+	 *            The path which the returned flows intersect.
+	 * @return Set of flows that cross at least one server on the path.
+	 * @throws Exception
+	 *			   Could not sort although there may be flows.
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Not sure where this description comes from. I checked all other occurrences, it does not seem helpful anywhere. In particular as there are not particular exceptions caught -- at the moment, the DNC's way to handle exceptions is to just hand them to the calling method until we reach the code that started the analysis, so the user can decide what to do with them.
Long story short, I suggest to remove this line and all other occurrences of "Could not sort although there may be flows."

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 2]
```diff
@@ -741,6 +741,14 @@ public Flow getFlow(int id) throws Exception {
 		}
 	}
 
+	/**
+	 * Finds all the flows that cross at least one server on the path p.
+	 * @param p
+	 *            The path which the returned flows intersect.
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: "The path *on* which ..."

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 3]
```diff
@@ -741,6 +741,14 @@ public Flow getFlow(int id) throws Exception {
 		}
 	}
 
+	/**
+	 * Finds all the flows that cross at least one server on the path p.
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Formatting: Please add an empty line between method and parameter description.

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 4]
```diff
@@ -757,6 +757,33 @@ public Flow getFlow(int id) throws Exception {
 		return result;
 	}
 
+    /**
+     * Finds all the flows that have exactly path p as their path.
+     * @param p
+     *             The path which all the returned flows have.
+     * @return Set of flows that have p as their path.
+     * @throws Exception
+     * 			   	Could not find although there may be flows.
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Remove this description of the exception (as with other methods)

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 5]
```diff
@@ -757,6 +757,33 @@ public Flow getFlow(int id) throws Exception {
...
+     */
+    public Set<Flow> getFlowsByPath(Path p) throws Exception {
+
+        Set<Flow> result = new HashSet<Flow>();;
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: two semi-columns at the end of the line

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 6]
```diff
@@ -757,6 +757,33 @@ public Flow getFlow(int id) throws Exception {
...
+
+        if(p != null )
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: else?
There is no flow having a null path, i.e., the missing else-clause could simply return the empty results set. However, that might lead to seemingly unexpected behavior that is very hard to debug. I prefer to throw a RuntimeException that breaks the entire execution.

Comment (alexscheffler) [reply src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: As it is right now, if the Path is null then the method will return an empty HashSet. Okay, I will throw a RuntimeException in that case instead.

Comment (sbondorf) [reply src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Can you add the according description to the JavaDocs, please?

**Comment by sbondorf** [comment]
I am ok with merging both changes -- javadocs for getFlows and the new getFlowsByPath method -- in a single pull request this time. But I'd prefer separate ones in the future, in particular if there are more complex commits. That should make it easier to review and discuss the PRs.

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java (hunk 6 above)
Comment (alexscheffler) [reply src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: I have added a comment within the method and in the description of the method for the case where the path is null.

**Comment by alexscheffler** [comment]
> Can you move the argument's null check to the very front of the method, please? Reasons are that we avoid creating the HashSet instance and (in my opinion) that simplifies the code by reducing the nesting depths.
> 
> Additionally, can you add a message to the new exception, e.g., "Path argument for method getFlowsByPath(Path p) was null." ? everything else of interest should be accessible by the exceptions trace / call stack.

Ok, done.

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 7]
```diff
@@ -749,6 +757,38 @@ public Flow getFlow(int id) throws Exception {
...
+	 */
+	public Set<Flow> getFlowsByPath(Path p) throws Exception {
+
+        // throw a runtime exception if p is null
+	    if(p == null)
+        {
+            throw new RuntimeException("Path argument for method getFlowsByPath(Path p) in ServerGraph was null.");
+        }
+
+	    else{
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Superfluous else after the if-clause throws a runtime exception.

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 8]
```diff
@@ -749,6 +757,38 @@ public Flow getFlow(int id) throws Exception {
...
+	 * @param p
+	 *             The path which all the returned flows have.
+	 * @return Set of flows that have p as their path.
+	 * @throws Exception
+	 *             If the path is null.
+	 */
+	public Set<Flow> getFlowsByPath(Path p) throws Exception {
+
+        // throw a runtime exception if p is null
+	    if(p == null)
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Please fix indentation.

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 9]
```diff
@@ -749,6 +757,38 @@ public Flow getFlow(int id) throws Exception {
...
+	 * Finds all the flows that have exactly path p as their path.
+	 *
+	 * @param p
+	 *             The path which all the returned flows have.
+	 * @return Set of flows that have p as their path.
+	 * @throws Exception
+	 *             If the path is null.
+	 */
+	public Set<Flow> getFlowsByPath(Path p) throws Exception {
+
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Remove empty line.

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 10]
```diff
@@ -749,6 +757,38 @@ public Flow getFlow(int id) throws Exception {
...
+        // throw a runtime exception if p is null
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: As this is a full sentence, please capitalize the first word and end it with a full stop.

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 11]
```diff
@@ -749,6 +757,38 @@ public Flow getFlow(int id) throws Exception {
...
+	    if(p == null)
+        {
+            throw new RuntimeException("Path argument for method getFlowsByPath(Path p) in ServerGraph was null.");
+        }
+
+	    else{
+            Set<Flow> result = new HashSet<Flow>();
+            // all the flows that start at first server of the path p are potential candidates
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Apply full sentence formatting here, too.
Shouldn't there be a definite article before "first"?

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 12]
```diff
@@ -749,6 +757,38 @@ public Flow getFlow(int id) throws Exception {
...
+            for(Flow f : flows_source_p)
+            {
+                if(f.getPath().equals(p))
+                {
+                    result.add(f);
+                }
+            }
+            return result;
+        }
+	}
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Indentation seems wrong here, too. Probably a mix of tabs and spaces. There's no formatting rules for DNC code (yet) but consistency within methods is appreciated.

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java [hunk 13]
```diff
@@ -741,6 +741,14 @@ public Flow getFlow(int id) throws Exception {
 		}
 	}
 
+	/**
+	 * Finds all the flows that cross at least one server on the path p.
+	 *
+	 * @param p
+	 *            The path on which the returned flows intersect.
+	 * @return Set of flows that cross at least one server on the path.
+	 * @throws Exception
```
Comment (sbondorf) [inline src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Lacks description.
In contrast to the clean-up of outdated and possibly wrong comments in other methods, newly added javadocs should be complete.

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java (hunk 8 above)
Comment (alexscheffler) [reply src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Looks different in my IDE...

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java (hunk 12 above)
Comment (alexscheffler) [reply src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Looks normal in my IDE though

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java (hunk 13 above)
Comment (alexscheffler) [reply src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: Any suggestions?

**Comment by alexscheffler** [comment]
I do not know why the indentation looks different here on GitHub. In my IDE everything looks fine...

**Comment by alexscheffler** [comment]
Ok, now the indentation on GitHub looks normal again

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java (hunk 13 above)
Comment (sbondorf) [reply src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: I traced potential exceptions in the methods called in the body. I think we get exceptions based if those methods are called with an argument set to null.
Overall, this leads to too much effort for this PR, should be done in a separate work item.

File: src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java (hunk 12 above)
Comment (sbondorf) [reply src/main/java/org/networkcalculus/dnc/network/server_graph/ServerGraph.java]: An elusive observation, unfortunately. Hitting tab might result in about four spaces that, in turn, look like a single tab in the IDE.

**Pull request merged**